### ISSUE TITLE
Flicker with --replace

### ISSUE THREAD

**Issue by CyberShadow** [dunst-project/dunst#320]
In a bash shell, run:

```bash
r=0; while true ; do r=$(dunstify --replace=$r --printid "Hi $r - $RANDOM") ; sleep 1 ; done
```

Very often there is a visible flicker when the notification is replaced.

Would it be possible to eliminate it?

**Comment by hobarrera** [comment]
Absolutely no flicker here. What WM are you using?

**Comment by CyberShadow** [comment]
i3-gaps. Can still see the flicker both on my desktop and laptop, with or without compositing (Compton).

**Comment by bebehei** [comment]
I can verify this. Using i3 with compton.

**Comment by lathan** [comment]
Same here, i3, different kind of flickering with or without compton. It happens with the volume notification from clementine, if you scroll the volume up/down fast enough. The flickering is slightly different with and without compton. With compton, sometimes the notification completely disappears until I move it to my other monitor. I never bothered reporting it because I have no idea where to start investigating this. :S

**Comment by gregf** [comment]
I can reproduce the flicker with the above script. i3-gaps no compton.

**Comment by gregf** [comment]
I have an amd video card using the opensource drivers.

**Comment by tsipinakis** [comment]
I can confirm that it also happens on i3wm, strangely enough it goes away if I disable compton.

**Comment by Deomkan** [comment]
I have bspwm with compton enabled and the graphic card is intel, and I got the flicker too

**Comment by DoktorL** [comment]
The notifications of the rhythmbox music player flicker as well ( i also use i3 gaps with compton).
After running dunst with -print i noticed that rhythmbox tries to print its notification multiple times for some reason, so i guess dunst flickers on i3/compton when it refreshes an already existing notification.

**Comment by josduj** [comment]
Setting `glx-use-copysubbuffermesa` to true fixed the compton flicker for me (only flickers slightly if I spam it like a maniac).

**Comment by reuseman** [comment]
@josduj Thank you for the solution, but there are still two problems.
1. If someone use the **xrender** backend
2. If someone use dunst to display a notification everytime the volume, brightness, etc., is increased or decreased. Or just something that needs high refresh, an that means spam like a maniac

**Comment by g-suraj** [comment]
Have the same issue, i3-gaps with compton.

**Comment by bebehei** [comment]
I guess, I've got the issue:

1. [notification.c:431](https://github.com/dunst-project/dunst/blob/a64cb829460fbd48d596fe9bf69653726c265db3/src/notification.c#L431)
    - The previous notification will get closed
1. [notification.c:574](https://github.com/dunst-project/dunst/blob/a64cb829460fbd48d596fe9bf69653726c265db3/src/notification.c#L574)
    - It will get removed from the displayed queue
1. [notification.c:596](https://github.com/dunst-project/dunst/blob/a64cb829460fbd48d596fe9bf69653726c265db3/src/notification.c#L596)
    - wake_up() will get called
1. [dunst.c:239](https://github.com/dunst-project/dunst/blob/a64cb829460fbd48d596fe9bf69653726c265db3/src/dunst.c#L239)
    - wich triggers the x window to hide
1. [notification.c:519](https://github.com/dunst-project/dunst/blob/a64cb829460fbd48d596fe9bf69653726c265db3/src/notification.c#L519)
    - bu until this point the notification queue is empty

So there is a full `x_win_show` and `x_win_hide` call in between replacing the notification. I'm not that experienced with X, but I'd guess this is the problem.

**Comment by bebehei** [comment]
If some of you could verify, that the fix proposed in #391 works, would be great.

**Comment by CyberShadow** [comment]
@bebehei Yes, it works for me! Can't speak about the patch but it does completely remove the flicker, both in the repro command above and my real use-case (showing current volume when adjusting it).
